EIPS/eip-6051.md

```diff
@@ -0,0 +1,469 @@
+
+- Raw bytes are encoded in hex and prefixed with '0x'.
+- Unless specified otherwise, all parameters and return values are hex-encoded bytes.
+- `cipher`  is encoded into single byte buffer as: `[IV  || encrypted_sk || tag]`.
+- `R`, `S`, `signerPubKey` and `trustedPubKey` are compressed if applicable.
+- `R` or `signerPubKey` could be followed by a signature to it: `[pub || sig]`. Note that for the secp256k1 curve, the signature is just 64 bytes without the `v` indicator as found in a typical Ethereum signature.
+
+#### R1. Request for Recipient to generate ephemeral key pair
+
+```javascript
+request({
+	method: 'eth_generateEphemeralKeyPair',
+	params: [version, signerPubKey],
+})
+// expected return value: R
+```
+
+`signerPubKey`  is optional. If provided, it is assumed that the implementation has the corresponding private key and the implementation MUST sign the ephemeral public key (in the form of what is to be returned). The signature algorithm is determined by the curve part of the  `version`  parameter, that is, ECDSA for secp256k1, and Ed25519 for Curve25519. And in this situation, it should be the case that the sender trusts  `signerPubKey`, no matter how this trust is maintained. If not, next request WILL be rejected by Sender application. Also see **Security Considerations**.
+
+Implementation then MUST generate random private key `r` with cryptographic secure random number generator (CSRNG), and derives ephemeral public key `R = [r]G`. Implementation SHOULD safe keep the generated key pair `(r, R)` in a secure manner in accordance with the circumstances, and SHOULD keep it only for limited duration, but the specific duration is left to individual implementations. Implementation SHOULD be able to retrieve `r` when given back the corresponding public key `R` if within said duration.
+
+Returned value is `R`, compressed if applicable. If  `signerPubKey`  is provided, then the `signature` is appended to `R`, also hex-encoded.
+
+Alternatively, `signature` could be calculated separately, and then appended to the returned data.
+
+#### R2. Request for Sender to encapsulate the private key
+
+```javascript
+request({
+	method: 'eth_encapsulatePrivateKey',
+	params: [
+		version,
+		recipient, // public key, may be followed by its signature, see signerPubKey
+		signerPubKey,
+		oob,
+		salt,
+		account
+	],
+})
+// expected return value: S || cipher
+```
+
+`recipient`  is the return value from the call to generate ephemeral key pair, with optional `signature` appended either as returned or separately.
+
+`oob`  and  `salt`  are just byte arrays.
+
+`account`  is used to identify which private key to be encapsulated. With Ethereum it is an address. Also see **Encoding of data and messages**.
+
+If  `signerPubKey`  is provided or  `recipient`  contains `signature` data, the implementation MUST perform signature verification. Missing data or incorrect format etc. SHALL fail the call and result in empty return and optional error logs.
```

Comment:
```suggestion
If  `signerPubKey`  is provided or  `recipient`  contains `signature` data, the implementation MUST perform signature verification. Missing data or incorrect format etc. SHALL fail the call and result in empty return and optional error logs.
```
```suggestion
If  `signerPubKey`  is provided or  `recipient`  contains `signature` data, the implementation MUST perform signature verification. Missing data or incorrect format MUST either fail the call or result in empty return and optional error logs.
```

I am not sure if "SHALL fail the call or return error with error logs" is logically sound here. Changing "MUST either fail the call, or return empty with error logs"

Reply:
Good catch.